crates/assembly-syntax/src/ast/item/resolver.rs

```diff
@@ -0,0 +1,390 @@
+use alloc::{collections::BTreeMap, sync::Arc, vec::Vec};
+
+use miden_debug_types::{SourceSpan, Span, Spanned};
+
+use crate::{
+    Path, Word,
+    ast::{AliasTarget, Ident, ItemIndex},
+    diagnostics::{Diagnostic, miette},
+};
+
+use super::{GlobalItemIndex, ModuleIndex};
+
+/// Represents the result of resolving a symbol
+#[derive(Debug, Clone)]
+pub enum SymbolResolution {
+    /// The name was resolved to a definition in the same module at the given index
+    Local(Span<ItemIndex>),
+    /// The name was resolved to an path referring to an item exported from another module
+    External(Span<Arc<Path>>),
+    /// The name was resolved to a procedure MAST root
+    MastRoot(Span<Word>),
+    /// The name was resolved to a known definition in the same module at the given index
+    Exact {
+        gid: GlobalItemIndex,
+        path: Span<Arc<Path>>,
+    },
+    /// The name was resolved to a known module
+    Module { id: ModuleIndex, path: Span<Arc<Path>> },
+}
+
+impl SymbolResolution {
+    pub fn into_global_id(&self) -> Option<GlobalItemIndex> {
+        match self {
+            Self::Exact { gid, .. } => Some(*gid),
+            Self::Local(_) | Self::External(_) | Self::MastRoot(_) | Self::Module { .. } => None,
+        }
+    }
+}
+
+impl Spanned for SymbolResolution {
+    fn span(&self) -> SourceSpan {
+        match self {
+            Self::Local(p) => p.span(),
+            Self::External(p) => p.span(),
+            Self::MastRoot(p) => p.span(),
+            Self::Exact { path, .. } => path.span(),
+            Self::Module { path, .. } => path.span(),
+        }
+    }
+}
+
+// LOCAL SYMBOL RESOLVER
+// ================================================================================================
+
+/// A lookup table for item names in the context of some module
+pub struct LocalSymbolResolver {
+    symbols: SymbolTable,
+}
+
+impl From<&crate::ast::Module> for LocalSymbolResolver {
+    fn from(module: &crate::ast::Module) -> Self {
+        Self { symbols: SymbolTable::from(module) }
+    }
+}
+
+impl From<&crate::library::ModuleInfo> for LocalSymbolResolver {
+    fn from(module: &crate::library::ModuleInfo) -> Self {
+        Self { symbols: SymbolTable::from(module) }
+    }
+}
+
+struct SymbolTable {
+    symbols: BTreeMap<Arc<str>, ItemIndex>,
+    imports: BTreeMap<Arc<str>, Result<SymbolResolution, LocalSymbolResolutionError>>,
+    items: Vec<Symbol>,
+}
+
+#[derive(Debug)]
+enum Symbol {
+    Item { name: Ident, resolved: SymbolResolution },
+    Import(Span<Arc<str>>),
+}
```

Comment:
Why do we need the `Import` variable? Isn't it equivalent to `SymbolResolution::External`, and hence could be encoded in the `Symbol::Item` variant? Might be worth documenting in a docstring